### ISSUE TITLE
Federating datasets doesn't work with subsets

### ISSUE THREAD

**Issue by schimpffabian** [OpenMined/PySyft#2071]
I don't know if this is an issue of general interest but PySyft would be more versatile if it were more agnostic in the way it handles datasets in dataset.federate((...))

I wrote a little tool to convert Subsets (also of ImageFolder datasets) to regular Datasets as a fix but it ain't a pretty solution:

```
class DatasetFromSubset(Dataset):
    def __init__(self, subset):
        data, targets = self.subset_to_dataset(subset)
        self.data = data
        self.targets = targets

    def __len__(self):
        return len(self.targets)

    def __getitem__(self, index):
        return self.data[index, :], self.targets[index]

    @staticmethod
    def subset_to_dataset(subset):
        indices = subset.indices
        targets = subset.dataset.targets

        targets_subset = torch.tensor(
            [targets[ii] for ii in list(indices.data.numpy())]
        )

        # Empty definition
        concat_tensor = None

        dataloader = torch.utils.data.DataLoader(subset, batch_size=2000, shuffle=False)

        for ii, (data, target) in enumerate(dataloader):
            print(ii)

            if ii == 0:
                concat_tensor = data
            else:
                concat_tensor = torch.cat((concat_tensor, data), 0)

return concat_tensor, targets_subset
```

**Comment by iamtrask** [comment]
Very interesting feature! What kind of subset are you interested in selecting?

**Comment by LaRiffle** [comment]
@iamtrask I assume it is linked to this example: https://github.com/2fasc/Distributed_Malaria/blob/master/src/federated_training.py

**Comment by 2fasc** [comment]
>  What kind of subset are you interested in selecting?
 I'm using `torch.utils.data.random_split(dataset, size_set)` to split datasets into train, test and validation sets. This basically creates a new class that contains the original dataset and adds a list of indices assigned to each split. 


> @iamtrask I assume it is linked to this example: https://github.com/2fasc/Distributed_Malaria/blob/master/src/federated_training.py

Exactly. I'm curious how imbalanced label distributions affect federated / distributed training.

**Comment by ewotawa** [comment]
@2fasc thanks for posting your code. How do I use it to convert an imagefolder dataset to a dataset that I can federate?
